numbat/src/name_resolution.rs

```diff
@@ -1,18 +1,76 @@
+use std::collections::HashMap;
+
 use thiserror::Error;
 
 use crate::span::Span;
 
 pub const LAST_RESULT_IDENTIFIERS: &[&str] = &["ans", "_"];
 
-#[derive(Debug, Clone, Error)]
+#[derive(Debug, Clone, Error, PartialEq, Eq)]
 pub enum NameResolutionError {
-    #[error("Identifier is already in use: '{conflicting_identifier}'.")]
+    #[error("Identifier is already in use{}: '{conflicting_identifier}'.",
+            if let Some(t) = .original_item_type { format!(" by the {t}") } else { "".to_owned() })]
     IdentifierClash {
         conflicting_identifier: String,
         conflict_span: Span,
         original_span: Span,
+        original_item_type: Option<String>,
     },
 
     #[error("Reserved identifier")]
     ReservedIdentifier(Span),
 }
+
+#[derive(Debug, Clone, Default)]
+pub struct Namespace {
+    seen: HashMap<String, (String, Span)>,
+}
+
+impl Namespace {
+    pub fn add_allow_override(
+        &mut self,
+        name: String,
+        span: Span,
+        item_type: String,
+    ) -> Result<(), NameResolutionError> {
+        self.add_inner(name, span, item_type, true)
+    }
+
+    pub fn add(
+        &mut self,
+        name: String,
+        span: Span,
+        item_type: String,
+    ) -> Result<(), NameResolutionError> {
+        self.add_inner(name, span, item_type, false)
+    }
+
+    fn add_inner(
```

Comment:
What does "inner" mean in this context? A small doc comment (or different name) might help here.